### PR TITLE
add TCC model to fescar

### PR DESCRIPTION
**Pull request by zhangthen**
### Ⅰ. Describe what this PR did

Add TCC model of Ant-Financial to fescar, the TCC model support local-beans, dubbo-bean,  and sofarpc-bean;

### CODE REVIEW

File: common/src/main/java/com/alibaba/fescar/common/Constants.java [hunk 1]
```diff
@@ -43,5 +43,56 @@
      */
     public static final String DBKEYS_SPLIT_CHAR = ",";
 
+    /** the start time of transaction  */
+    public static final String START_TIME  = "start-time";
+    
+    /**
+     * 应用名称
```
Comment (jovany-wang) [inline common/src/main/java/com/alibaba/fescar/common/Constants.java]: Let's comment in English.

File: common/src/main/java/com/alibaba/fescar/common/executor/Initialize.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package com.alibaba.fescar.common.executor;
+
+/**
+ * @author zhangsen
+ */
+public interface Initialize {
```
Comment (jovany-wang) [inline common/src/main/java/com/alibaba/fescar/common/executor/Initialize.java]: Could you please rename this method?

File: common/src/main/java/com/alibaba/fescar/common/loader/EnhancedServiceLoader.java [hunk 3]
```diff
@@ -20,14 +20,11 @@
 import java.io.IOException;
 import java.io.InputStreamReader;
 import java.net.URL;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.Comparator;
-import java.util.Enumeration;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (jovany-wang) [inline common/src/main/java/com/alibaba/fescar/common/loader/EnhancedServiceLoader.java]: Don't use `*` to import packages.

File: common/src/main/java/com/alibaba/fescar/common/Constants.java (hunk 1 above)
Comment (zhangthen) [reply common/src/main/java/com/alibaba/fescar/common/Constants.java]: Good suggestion.

File: common/src/main/java/com/alibaba/fescar/common/loader/EnhancedServiceLoader.java (hunk 3 above)
Comment (zhangthen) [reply common/src/main/java/com/alibaba/fescar/common/loader/EnhancedServiceLoader.java]: Idea do it automatically

File: common/src/main/java/com/alibaba/fescar/common/executor/Initialize.java (hunk 2 above)
Comment (zhangthen) [reply common/src/main/java/com/alibaba/fescar/common/executor/Initialize.java]: rename to what ?

Comment (khotyn) [reply common/src/main/java/com/alibaba/fescar/common/executor/Initialize.java]: @zhangthen Turn the first letter of the method name to lowercase.

Comment (zhangthen) [reply common/src/main/java/com/alibaba/fescar/common/executor/Initialize.java]: The  first letter of the  method ‘init’  which is the  only method of class  Initialize  always been   lowercase.

**Comment by slievrly** [comment]
@skyesx @github-ygy please review TCC mode

File: common/src/main/java/com/alibaba/fescar/common/util/StringUtils.java [hunk 4]
```diff
@@ -43,6 +43,44 @@ public static final boolean isEmpty(String str) {
         return (str == null) || (str.isEmpty());
     }
 
+    /**
+     * Is blank string ?
+     * @param str
+     * @return
+     */
+    public static boolean isBlank(String str) {
```
Comment (lovepoem) [inline common/src/main/java/com/alibaba/fescar/common/util/StringUtils.java]: How about use the `apache commons.lang.jar/lang3.jar` ?   The common jar also have the ` StringUtils.isBlank` method

Comment (lovepoem) [reply common/src/main/java/com/alibaba/fescar/common/util/StringUtils.java]: apache commons   also have `CollectionUtils.isBlank` method

File: common/src/main/java/com/alibaba/fescar/common/Constants.java [hunk 5]
```diff
@@ -43,5 +43,56 @@
      */
     public static final String DBKEYS_SPLIT_CHAR = ",";
 
+    /** the start time of transaction  */
+    public static final String START_TIME  = "start-time";
```
Comment (ranqiqiang) [inline common/src/main/java/com/alibaba/fescar/common/Constants.java]: "-"和"_"  我看都有在用，默认约定吗？

File: common/src/main/java/com/alibaba/fescar/common/loader/EnhancedServiceLoader.java [hunk 6]
```diff
@@ -102,6 +104,28 @@
...
+    /**
+     * 返回所有实现类实例
+     * @param service
+     * @param <S>
+     * @return
+     */
+    public static <S> List<S> loadAll(Class<S> service){
+        List<S> allInstances = new ArrayList<>();
+        List<Class> allClazzs = getAllExtensionClass(service);
+        if(CollectionUtils.isEmpty(allClazzs)){
```
Comment (ranqiqiang) [inline common/src/main/java/com/alibaba/fescar/common/loader/EnhancedServiceLoader.java]: 这里确定是为空就返回？

Comment (ranqiqiang) [reply common/src/main/java/com/alibaba/fescar/common/loader/EnhancedServiceLoader.java]: 好吧，这个后面我看到代码了- -，但是感觉有点反常规啊，这个命名
org.apache.commons.collections.CollectionUtils
也有这个方法..

File: core/src/main/java/com/alibaba/fescar/core/protocol/transaction/BranchRegisterRequest.java [hunk 7]
```diff
@@ -114,15 +116,34 @@ public short getTypeCode() {
...
+
     @Override
     public byte[] encode() {
+        int byteLenth = 0;
         byte[] lockKeyBytes = null;
         if (this.lockKey != null) {
             lockKeyBytes = lockKey.getBytes(UTF8);
             if (lockKeyBytes.length > 512) {
-                byteBuffer = ByteBuffer.allocate(lockKeyBytes.length + 1024);
+//                byteBuffer = ByteBuffer.allocate(lockKeyBytes.length + 1024);
```
Comment (ranqiqiang) [inline core/src/main/java/com/alibaba/fescar/core/protocol/transaction/BranchRegisterRequest.java]: 注释的可以删除，变量可以定义出去，后面可以支持参数

File: common/src/main/java/com/alibaba/fescar/common/util/StringUtils.java (hunk 4 above)
Comment (zhangthen) [reply common/src/main/java/com/alibaba/fescar/common/util/StringUtils.java]: @lovepoem I think it's better to dependent as little as possible on the third-party packages, if the apache jar is not used, we can exclude it.

File: common/src/main/java/com/alibaba/fescar/common/Constants.java (hunk 5 above)
Comment (zhangthen) [reply common/src/main/java/com/alibaba/fescar/common/Constants.java]: @ranqiqiang  I had  fixed it.

File: common/src/main/java/com/alibaba/fescar/common/loader/EnhancedServiceLoader.java (hunk 6 above)
Comment (zhangthen) [reply common/src/main/java/com/alibaba/fescar/common/loader/EnhancedServiceLoader.java]: @ranqiqiang  I think it's better to dependent as little as possible on the third-party packages, If  the Apache jar can be replaced, we will exclude it.

File: core/src/main/java/com/alibaba/fescar/core/protocol/transaction/BranchRegisterRequest.java (hunk 7 above)
Comment (zhangthen) [reply core/src/main/java/com/alibaba/fescar/core/protocol/transaction/BranchRegisterRequest.java]: @ranqiqiang  The notes had been removed, the variables can be defined when we add parameters.

**Comment by skyesx** [comment]
this is a huge PR，  if author can summarize  the design principles and major modifications  will be easy for us to  understand.

and also i think this PR is already have a lot of conflicts with develop branch...

**Comment by slievrly** [comment]
Temporarily merged into feature, re-pr, convenient for everyone to review

**Pull request merged**